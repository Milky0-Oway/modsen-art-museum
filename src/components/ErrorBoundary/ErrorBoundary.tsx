import React, { Component, ErrorInfo } from 'react';

type Props = {
	children: React.ReactNode;
};

type State = {
	hasError: boolean;
};

class ErrorBoundary extends Component<Props, State> {
	constructor(props: Props) {
		super(props);
		this.state = { hasError: false };
	}

	static getDerivedStateFromError(error: Error) {
		console.log(error);
		return { hasError: true };
	}

	componentDidCatch(error: Error, errorInfo: ErrorInfo) {
		console.error('Error in ErrorBoundary: ', error, errorInfo);
	}

	render() {
		if (this.state.hasError) {
			return <h1>Something went wrong...</h1>;
		}

		return this.props.children;
	}
}

export default ErrorBoundary;
