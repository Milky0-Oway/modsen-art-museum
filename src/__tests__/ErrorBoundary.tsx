import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import ErrorBoundary from 'components/ErrorBoundary/ErrorBoundary';

const ProblematicComponent = () => {
	throw new Error('I crashed!');
};

describe('ErrorBoundary', () => {
	it('should render its children when there is no error', () => {
		const { container } = render(
			<ErrorBoundary>
				<h1>All is well</h1>
			</ErrorBoundary>,
		);

		expect(screen.getByText('All is well')).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it('should catch an error and render fallback UI', () => {
		const { container } = render(
			<ErrorBoundary>
				<ProblematicComponent />
			</ErrorBoundary>,
		);

		expect(screen.getByText('Something went wrong...')).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});
});
