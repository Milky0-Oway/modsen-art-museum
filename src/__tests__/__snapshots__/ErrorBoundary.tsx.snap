// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should catch an error and render fallback UI 1`] = `
<div>
  <h1>
    Something went wrong...
  </h1>
</div>
`;

exports[`ErrorBoundary should render its children when there is no error 1`] = `
<div>
  <h1>
    All is well
  </h1>
</div>
`;
